compiler/src/dotty/tools/dotc/core/JavaNullInterop.scala

```diff
@@ -56,94 +56,35 @@ object JavaNullInterop {
-  private case class MethodPolicy(trigger: Symbol => Boolean,
-                                  nnParams: Seq[Int],
-                                  nnRes: Boolean)(implicit ctx: Context) extends TypeMap with NullifyPolicy {
-    override def isApplicable(sym: Symbol): Boolean = trigger(sym)
-
-    private def spare(tp: Type): Type = {
-      nullifyType(tp).stripNull
-    }
-
-    override def apply(tp: Type): Type = {
-      tp match {
-        case ptp: PolyType =>
-          derivedLambdaType(ptp)(ptp.paramInfos, this(ptp.resType))
-        case mtp: MethodType =>
-          val paramTpes = mtp.paramInfos.zipWithIndex.map {
-            case (paramInfo, index) =>
-              // TODO(abeln): the sequence lookup can be optimized, because the indices
-              // in it appear in increasing order.
-              if (nnParams.contains(index)) spare(paramInfo) else nullifyType(paramInfo)
-          }
-          val resTpe = if (nnRes) spare(mtp.resType) else nullifyType(mtp.resType)
-          derivedLambdaType(mtp)(paramTpes, resTpe)
-      }
-    }
+    // Some special cases when nullifying the type
+    if (sym.name == nme.TYPE_ || sym.isAllOf(Flags.JavaEnumValue))
+      // Don't nullify the `TYPE` field in every class and Java enum instances
+      tp
+    else if (sym.name == nme.toString_ || sym.isConstructor)
+      // Don't nullify the return type of the `toString` method and constructors
+      nullifyParamsOnly(tp)
+    else
+      // Otherwise, nullify everything
+      nullifyType(tp)
   }
 
-  /** A policy that nullifies only method parameters (but not result types). */
-  private def paramsOnlyPolicy(trigger: Symbol => Boolean)(implicit ctx: Context): MethodPolicy = {
-    MethodPolicy(trigger, nnParams = Seq.empty, nnRes = true)
-  }
+  /** Only nullify method parameters (but not result types). */
+  private def nullifyParamsOnly(tp: Type)(implicit ctx: Context): Type =
+    new JavaNullMap(alreadyNullable = false, nnRes = true)(ctx)(tp)
 
   /** Nullifies a Java type by adding `| JavaNull` in the relevant places. */
-  private def nullifyType(tpe: Type)(implicit ctx: Context): Type = {
-    val nullMap = new JavaNullMap(alreadyNullable = false)
-    nullMap(tpe)
-  }
+  private def nullifyType(tp: Type)(implicit ctx: Context): Type =
+    new JavaNullMap(alreadyNullable = false, nnRes = false)(ctx)(tp)
 
   /** A type map that adds `| JavaNull`.
    *  @param alreadyNullable whether the type being mapped is already nullable (at the outermost level).
    *                         This is needed so that `JavaNullMap(A | B)` gives back `(A | B) | JavaNull`,
    *                         instead of `(A|JavaNull | B|JavaNull) | JavaNull`.
+   *  @param nnRes whether the result type should be nullified.
    */
-  private class JavaNullMap(var alreadyNullable: Boolean)(implicit ctx: Context) extends TypeMap {
+  private class JavaNullMap(var alreadyNullable: Boolean, nnRes: Boolean)(implicit ctx: Context) extends TypeMap {
```

Comment:
Could you please rename `nnRes` to something like `nonNullResultType` that's a bit more legible?
(I realize `nnRes` is what I had before :) )

Reply:
Also, I think `nnRes` should probably be a `var`. If we somehow end with nested MethodTypes in the return type, we really only want to not nullify the outermost occurrence. `nnRes` would need to be set to false before the recursive calls in the MethodType case.